docs/HighLevelDesignForGraphProgramExtraction.md

```diff
@@ -0,0 +1,70 @@
+# Graph Program Extraction: High-Level Design Proposal
+
+This document aims to serve as complementary to the original graph program extraction (GPE) [whitepaper](<https://github.com/tensorflow/swift/blob/master/docs/GraphProgramExtraction.md>). It is a proposal for a general high-level design for performing GPE. Even though the document uses Swift and TensorFlow as the running use case, the proposed design is more general and could be used with other languages and frameworks. It is also a very early draft but the hope is that it will stimulate discussion and evolve into a few well-defined and well-specified design guidelines.
+
+## Desired Features
+
+As described in the GPE whitepaper, GPE is effectively a [program slicing](https://en.wikipedia.org/wiki/Program_slicing) method. This can be thought of as follows: working with Swift and TensorFlow means that we work with two different languages that get executed as two separate programs that can communicate when needed. One program is the main Swift program that also serves as the entrypoint of execution. The other is the _implicitly constructed_ TensorFlow graph that gets executed by the TensorFlow runtime. Ideally we would like to:
+
+1. Have a single entrypoint for the Swift program and a single entrypoint for the TensorFlow program (i.e., use a single TF graph for the whole program, and a single session that starts executing parts of the graph at the same time that the Swift program starts executing things on the Swift-side). _This is because creating multiple sessions can be expensive and also because resource sharing across sessions may not be easy to control (I am not 100% sure about this second comment, but it makes sense to me intuitively)_.
```

Comment:
Agree this would be a really clean design! My immediate next step is to generate a single graph for each Swift module (but still retain the start graph / stop graph calls for each partitionable function). That should help us march towards the direction described here.

Reply:
That sounds like a good first step that I think can start simplifying things. :)